### PR TITLE
Fixes Issue 1614

### PR DESCRIPTION
**Pull request by greglandrum**
This is a replacement for #1626.

### CODE REVIEW

**Comment by bp-kelley** [comment]
@greglandrum Starting to look at it now...

File: Code/GraphMol/testChirality.cpp [hunk 1]
```diff
@@ -631,6 +632,274 @@ void testChiralityCleanup() {
...
+// CURRENTLY WORK
+// - bond stereo dependent on atom stereo breaking symmetry - DOESN'T CURRENTLY
+// WORK
+// - atom stereo dependent on other atom stereo breaking symmetry - WORKS
+// - atom stereo dependent on bond stereo breaking symmetry - WORKS
+///////////////////////////
+
+// Everything ifdef'd USE_NEW_STEREOCHEMISTRY are test cases that
+// should 'hopefully' work when switching to the new_canon.h ranking
+// for stereo perception. However, making USE_NEW_STEREOCHEMISTRY
```
Comment (bp-kelley) [inline Code/GraphMol/testChirality.cpp]: I'm curious what this means. Is this saying that the morgan style canonicalization for R/S E/Z is incompatible with the current implementation?

You know my feelings about R/S E/Z identification :) I'm tempted to make a PR to have the substructure search not require it, but that's a lot of work!

File: Code/GraphMol/Wrap/MolOps.cpp [hunk 2]
```diff
@@ -1544,7 +1544,9 @@ struct molops_wrapper {
     - mol: the molecule to use\n\
     - cleanIt: (optional) if provided, atoms with a chiral specifier that aren't\n\
       actually chiral (e.g. atoms with duplicate substituents or only 2 substituents,\n\
-      etc.) will have their chiral code set to CHI_UNSPECIFIED\n\
+      etc.) will have their chiral code set to CHI_UNSPECIFIED. Bonds with \n\
+      STEREOCIS/STEREOTRANS specified that have duplicate substituents based upon the CIP \n\
+      atom ranks will be marked STEREONONE. \n\
```
Comment (bp-kelley) [inline Code/GraphMol/Wrap/MolOps.cpp]: Ah, I know see why CIP is important here.

Comment (greglandrum) [reply Code/GraphMol/Wrap/MolOps.cpp]: yeah, until the morgan-based approach is done we're unfortunately stuck with it.

**Comment by bp-kelley** [comment]
Finally wrapped my head around this.

**Comment by bp-kelley** [comment]
Do you want to change the \param I mentioned?  It’s not important overall but I find the current doc  a bit confusing.

**Comment by greglandrum** [comment]
Yep, sorry, forgot to do that. Done now.

**Comment by greglandrum** [comment]
@coleb @bp-kelley : I'm going to merge this tomorrow (Monday) morning unless you have objections.

**Comment by bp-kelley** [comment]
Looks good to me

File: Code/GraphMol/MolOps.h [hunk 3]
```diff
@@ -820,14 +821,23 @@ void assignStereochemistryFrom3D(ROMol &mol, int confId = -1,
...
-  \param mol     the molecule of interest
-  \param cleanIt toggles removal of stereo flags from double bonds that can
-                 not have stereochemistry
-  \param force   forces the calculation to be repeated even if it has
+  Does the CIP stereochemistry assignment for the molecule's atoms
+  (R/S) and double bond (Z/E). Chiral atoms will have a property
+  '_CIPCode' indicating their chiral code.
+
+  \param mol     the molecule to use
+  \param cleanIt if true, atoms with a chiral specifier that aren't
```
Comment (coleb) [inline Code/GraphMol/MolOps.h]: I don't think this documentation matches what the function does. This function appears to only affect bond directions and not atoms at all.

File: Code/GraphMol/testChirality.cpp [hunk 4]
```diff
@@ -631,6 +632,274 @@ void testChiralityCleanup() {
   TEST_ASSERT(mol->getAtomWithIdx(1)->getChiralTag() == Atom::CHI_UNSPECIFIED);
   delete mol;
 
+  // The remaining examples are for github #1614 :
+  // AssignStereochemistry incorrectly removing CIS/TRANS bond stereo
+
+  // cleanIt=true, force=true should NOT remove this manual cis/trans stereo
+  // assignment
+  smi = "CC=CC(Cl)C";
```
Comment (coleb) [inline Code/GraphMol/testChirality.cpp]: Just manually reviewed all the SMILES in this test case. All the assertions look correct to me.

File: Code/GraphMol/molopstest.cpp [hunk 5]
```diff
@@ -6978,6 +6978,222 @@ void testGithub1703() {
...
+    m.addBond(2, 3, Bond::SINGLE);
+    m.addBond(1, 2, Bond::DOUBLE);
+    m.getBondBetweenAtoms(1, 2)->setStereoAtoms(0, 3);
+    m.getBondBetweenAtoms(1, 2)->setStereo(Bond::STEREOTRANS);
+    m.updatePropertyCache();
+
+    {
+      RWMol nm(m);
+      bool force = true, cleanIt = true;
+      MolOps::setDoubleBondNeighborDirections(nm);
```
Comment (coleb) [inline Code/GraphMol/molopstest.cpp]: It looks like we'll be requiring that setDoubleBondNeighborDirections be called after manually setting Bond::setStereo(STEREOCIS/STEREOTRANS)? That seems fine as the function name is quite explicitly named, but I'm curious about why this design is necessary?

@greglandrum, did you try folding MolOps::setDoubleBondNeighborDirections into the preamble of MolOps::assignStereochemistry? Would this be a bad idea?

**Comment by coleb** [comment]
In general, everything looks good to me. My comments are fairly minor. I originally ran into this bug while trying to extend EnumerateStereoisomers for pseudo-stereo centers. It would be great if pseudo-stereo double bonds were handled as indicated in the test comment for completeness, but pseudo-stereo atom centers I believe are more common in practice, and those do seem to be handled properly. I'll continue that work based on this branch to see if I spot any other problems.

File: Code/GraphMol/MolOps.h (hunk 3 above)
Comment (greglandrum) [reply Code/GraphMol/MolOps.h]: just to make sure we're looking at the same function; this is `assignStereochemistry()`, which certainly operates on both atoms and bonds.

Comment (coleb) [reply Code/GraphMol/MolOps.h]: Oops! You're correct, the diff was making me think the doc-string came after the declaration of setDoubleBondNeighborDirections.

File: Code/GraphMol/molopstest.cpp (hunk 5 above)
Comment (greglandrum) [reply Code/GraphMol/molopstest.cpp]: It's not so much that it's a bad idea, but I (think) I've tried to keep the bond directions unset whenever possible. So double bond stereo is stored on the double bonds themselves, and not set redundantly on their neighboring bonds unless client code requests it.

Comment (coleb) [reply Code/GraphMol/molopstest.cpp]: That seems reasonable to never set bond directions until they're needed (really only the SMILES writer?). Are they a prerequisite for MolOps::assignStereochemistry now though?

Comment (greglandrum) [reply Code/GraphMol/molopstest.cpp]: hmm, looks like it is still necessary for assignStereochemistry.

You definitely don't need it to get correct SMILES:
```
In [13]: m = Chem.MolFromSmiles('CC=CC')

In [14]: m.GetBondWithIdx(1).SetStereoAtoms(0,3)

In [15]: m.GetBondWithIdx(1).SetStereo(Chem.BondStereo.STEREOCIS)

In [16]: Chem.MolToSmiles(m,True)
Out[16]: 'C/C=C\\C'
```

but to have AssignStereochemistry() work it's necessary:
```
In [24]: m = Chem.MolFromSmiles('CC=CC')

In [25]: m.GetBondWithIdx(1).SetStereoAtoms(0,3)

In [26]: m.GetBondWithIdx(1).SetStereo(Chem.BondStereo.STEREOCIS)

In [27]: Chem.AssignStereochemistry(m,cleanIt=True,force=True)

In [28]: m.GetBondWithIdx(1).GetStereo()
Out[28]: rdkit.Chem.rdchem.BondStereo.STEREONONE
```
That feels like a bug (though not necessarily one to hold up this merge). What's your take?

Comment (coleb) [reply Code/GraphMol/molopstest.cpp]: Not necessary to hold up this merge. I'll play with it in the context of expanding EnumerateStereoisomers for pseudo-stereo and let you know how it goes there.

**Pull request merged**